### PR TITLE
record level permissions: added DELETE endpoints

### PR DESCRIPTION
**Pull request by shriramshankar**
- added DELETE /v1/[subjects|aspects|perspectives|lenses]/{key}/writers
- added DELETE /v1/[subjects|aspects|perspectives|lenses]/{key}/writers/userNameOrId
- made a few changes to DELETE /v1/[subjects|aspects|perspectives|lenses|users]/{key}

### CODE REVIEW

**Comment by shriramshankar** [comment]
@iamigo @annyhe  This PR has all the changes for the delete endpoint. I will create another PR for the "403" status I am once back.

File: api/v1/helpers/verbs/doDelete.js [hunk 1]
```diff
@@ -32,7 +32,20 @@ function doDelete(req, res, next, props) {
     if (props.loggingEnabled) {
       logAPI(req, props.modelName, o);
     }
+    const assocNames = [];
 
+    /**
```
Comment (iamigo) [inline api/v1/helpers/verbs/doDelete.js]: Spacing/formatting/style should be...

```
    /*
     * If props.belongsToManyAssoc defined, take the values of the object and
     * push it into the assocNames array.
     */
    if (props.belongsToManyAssoc) {
      Object.keys(props.belongsToManyAssoc)
      .forEach((key) => assocNames.push(props.belongsToManyAssoc[key]));
    }
```

File: api/v1/controllers/aspects.js [hunk 2]
```diff
@@ -187,6 +191,34 @@ module.exports = {
...
+   *
+   * Deletes all the writers associated with this resource.
+   *
+   * @param {IncomingMessage} req - The request object
+   * @param {ServerResponse} res - The response object
+   * @param {Function} next - The next middleware function in the stack
+   */
+  deleteAspectWriters(req, res, next) {
+    doDeleteAllAssoc(req, res, next, helper, helper.belongsToManyAssoc.users);
+  },
```
Comment (annyhe) [inline api/v1/controllers/aspects.js]: @shriramshankar is there a reason why we need to pass both helper and helper.belongsToManyAssoc.users?

File: api/v1/controllers/aspects.js [hunk 3]
```diff
@@ -187,6 +191,34 @@ module.exports = {
...
+   * Deletes a user from an aspect’s list of authorized writers.
+   *
+   * @param {IncomingMessage} req - The request object
+   * @param {ServerResponse} res - The response object
+   * @param {Function} next - The next middleware function in the stack
+   */
+  deleteAspectWriter(req, res, next) {
+    const userNameOrId = req.swagger.params.userNameOrId.value;
+    doDeleteOneAssoc(req, res, next, helper,
+        helper.belongsToManyAssoc.users, userNameOrId);
```
Comment (annyhe) [inline api/v1/controllers/aspects.js]: @shriramshankar is there a reason why we need to pass both helper and helper.belongsToManyAssoc.users?

File: api/v1/controllers/lenses.js [hunk 4]
```diff
@@ -147,6 +151,34 @@ module.exports = {
...
+   * DELETE /lenses/{keys}/writers
+   *
+   * Deletes all the writers associated with this resource.
+   *
+   * @param {IncomingMessage} req - The request object
+   * @param {ServerResponse} res - The response object
+   * @param {Function} next - The next middleware function in the stack
+   */
+  deleteLensWriters(req, res, next) {
+    doDeleteAllAssoc(req, res, next, helper, helper.belongsToManyAssoc.users);
```
Comment (annyhe) [inline api/v1/controllers/lenses.js]: @shriramshankar same question as above

File: api/v1/helpers/verbs/doDeleteAllBToMAssoc.js [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+ * @param {ServerResponse} res - The response object
+ * @param {Function} next - The next middleware function in the stack
+ * @param {Module} props - The module containing the properties of the resource
+ * whose association has to be deleted
+ * @param {String} assocName - The name of the association associated with the
+ * model
+ *
+ */
+function doDeleteAllBToMAssoc(req, res, next, props, assocName) { // eslint-disable-line
+  const params = req.swagger.params;
```
Comment (annyhe) [inline api/v1/helpers/verbs/doDeleteAllBToMAssoc.js]: @shriramshankar can we get assocName from props?

**Comment by shriramshankar** [comment]
@iamigo @annyhe  PR updated.

File: api/v1/controllers/aspects.js (hunk 2 above)
Comment (shriramshankar) [inline api/v1/controllers/aspects.js]: all the doDelete*.js files are generic. They are created to delete associations between any two models, not just  user model. The passing helper.belongsToManyAssoc.users to doDeleteAllAssoc lets us delete just the user associations.

File: api/v1/helpers/verbs/doDeleteAllBToMAssoc.js (hunk 5 above)
Comment (shriramshankar) [inline api/v1/helpers/verbs/doDeleteAllBToMAssoc.js]: This is needed.

File: api/v1/helpers/verbs/doDelete.js (hunk 1 above)
Comment (shriramshankar) [inline api/v1/helpers/verbs/doDelete.js]: changes done.

**Comment by shriramshankar** [comment]
@annyhe Can you merge this branch if you are done with the review? This needs to be merged for my next PR.

**Comment by annyhe** [comment]
@shriramshankar @iamigo The tests are failing on travis due to "The command "npm run jscs-enforce" exited with 1."
Either clean up the jscs errors, or commit without running the command.

**Comment by iamigo** [comment]
we all need to get used to clearing jscs errors :(

**Comment by shriramshankar** [comment]
@iamigo @annyhe  the lint errors are fixed. Can we please merge this now?

**Pull request merged**